va/va_drmcommon.h

```diff
@@ -154,5 +154,20 @@ typedef struct _VADRMPRIMESurfaceDescriptor {
     } layers[4];
 } VADRMPRIMESurfaceDescriptor;
 
+/**
+ * \brief List of DRM format modifiers.
+ *
+ * To allocate surfaces with one of the modifiers specified in the array, call
+ * vaCreateSurfaces() with the VASurfaceAttribDRMFormatModifiers attribute set
+ * to point to an array of num_surfaces instances of this structure. The driver
+ * will select the optimal modifier in the list.
+ */
+typedef struct _VADRMFormatModifierList {
+	/** Number of modifiers. */
+	uint32_t num_modifiers;
+	/** Array of modifiers. */
+	uint64_t *modifiers;
+} VADRMFormatModifierList;
+
```

Comment:
Is this used for list all supported modifiers of some config? For VPP, we really think the query result
of VASurfaceAttribPixelFormat is both supported for input and output. Also for the modifier, if input and
output can support the same types of modifiers, we can still do the same way. But if not, we may need to
differentiate which modifiers are for input and which are for output.

Reply:
I see.

No, this is not used to list all supported modifiers of a given config. It's only used with `vaCreateSurfaces`.

Reply:
So, I think the ideal usage of this modifier should be:
1. query the supported modifiers list based on the configID. For example, an encoder
   configID on AMD platform, the query may contain linear and Y_TILE, but on Intel
   platform it may just contain Y_TILE.
2. Then, we select one supported modifier using this attribute to create surface.

For VPP, maybe the things are a little complicated, if input and output modifiers are
different.

Reply:
I don't think the libva client should be responsible for picking the modifier.

There may be multiple modifiers suitable for a given usage. For instance, both Y_TILED and X_TILED may be supported by KMS for a plane. However the client can't guess what modifier is "the best one".

The VA-API driver is a better place to decide which exact modifier to pick. The VA-API driver knows that Y_TILED is better than X_TILED.

That's what the GBM API does too: `gbm_bo_create_with_modifiers` takes a list of modifiers, and the driver picks the best one from the list. See https://gitlab.freedesktop.org/mesa/mesa/-/blob/f532202f2d55b9ac475b7e3f8c96a4dd23489299/src/gbm/main/gbm.h#L275.

Reply:
This sounds like a good plan to me.

However you may notice that the libva query is not strictly necessary. The libva client can just pass the list of modifiers suitable for the usage (e.g. the list of modifiers supported by EGL or KMS). The libva driver can internally intersect the user list with its own list of supported modifiers. That's why I left the query part out.

I'm still willing to add a query API nonetheless if you want it, but it's still unclear to me how to address the memory ownership issue. The driver needs to return something more complex than just a single integer to the user. `VAGenericValue` can hold a pointer, but who's responsible for allocating/free'ing the memory?

Reply:
As a gstreamer developer, the query is useful for negotiation. For example, if the
query result of supported modifiers of VPP output is LINEAR and Y_TILE, but the
3D module only support X_TILE input, we can quit in very early stage and report
the error reason more precise. This may also apply to other media frameworks such
as FFMpeg, etc.

About the "memory ownership issue", I am not the VA expert:)

Reply:
Right. By "not strictly necessary", I mean that it would be nice to have but would still work without it. You'd just get an error from `vaCreateSurfaces` if VPP output doesn't support EGL modifiers.

Thanks for the feedback!